core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -2271,36 +2299,39 @@ object Parser {
       }
     }
 
-    case class Rep0[A, B](p1: Parser[A], acc: Accumulator0[A, B]) extends Parser0[B] {
+    case class Rep0[A, B](p1: Parser[A], max: Int, acc: Accumulator0[A, B]) extends Parser0[B] {
       private[this] val ignore: B = null.asInstanceOf[B]
 
       override def parseMut(state: State): B = {
         if (state.capture) {
           val app = acc.newAppender()
-          if (repCapture(p1, 0, state, app)) app.finish()
+          if (repCapture(p1, 0, max, state, app)) app.finish()
           else ignore
         } else {
-          repNoCapture(p1, 0, state)
+          repNoCapture(p1, 0, max, state)
           ignore
         }
       }
     }
 
-    case class Rep[A, B](p1: Parser[A], min: Int, acc1: Accumulator[A, B]) extends Parser[B] {
+    /** A parser that can repeats the underlying parser multiple times
+      */
+    case class Rep[A, B](p1: Parser[A], min: Int, max: Int, acc1: Accumulator[A, B])
+        extends Parser[B] {
       if (min < 1) throw new IllegalArgumentException(s"expected min >= 1, found: $min")
 
       private[this] val ignore: B = null.asInstanceOf[B]
 
       override def parseMut(state: State): B = {
         val head = p1.parseMut(state)
-
+        def maxRemaining = if (max == Int.MaxValue) max else max - 1
```

Comment:
Feels like we are doubling this up. I'm confused how this is here and in the rep methods above. Can we do it in only one place?

Reply:
This is here because we pre-parse one for the appender in rep (but not rep0 or the non-capturing varieties) before we do things with loops, so we need to parse at most max - 1 in the loop unless max == Int.MaxValue.

In the loop we test for `cnt <= loopmax - 1` so that if we need to enter the loop no more than `loopmax` times unless `loopmax == Int.MaxValue`, in which case we can enter the loop more than Int.MaxValue times because it never evaluates to false.

This wasn't apparent in the variation with the sentinel being a negative value, since it tested for max < 0, and it didn't matter whether max ended up being -1 or -2.

I'll see if I can find a clearer alternative encoding.